python/cudf/cudf/core/resample.py

```diff
@@ -0,0 +1,360 @@
+# Copyright (c) 2021, NVIDIA CORPORATION.
+
+import numpy as np
+import pandas as pd
+
+import cudf
+from cudf._typing import DataFrameOrSeries
+from cudf.core.groupby.groupby import (
+    DataFrameGroupBy,
+    GroupBy,
+    Grouper,
+    SeriesGroupBy,
+    _Grouping,
+)
+
+
+class _Resampler(GroupBy):
+
+    grouping: "_ResampleGrouping"
+
+    def __init__(self, obj, by, axis=None, kind=None):
+        if not isinstance(by, Grouper) and by.freq:
+            # TODO
+            raise ValueError()
+
+        by = _ResampleGrouping(obj, by)
+        super().__init__(obj, by=by)
+
+    def agg(self, func):
+        result = super().agg(func)
+        if len(self.grouping.bin_labels) != len(result):
+            index = cudf.Index(
+                self.grouping.bin_labels, name=self.grouping.names[0]
+            )
+            return result._align_to_index(
+                index, how="right", sort=False, allow_non_unique=True
+            )
+        else:
+            return result.sort_index()
+
+    def asfreq(self):
+        return self.obj._align_to_index(
+            self.grouping.bin_labels,
+            how="right",
+            sort=False,
+            allow_non_unique=True,
+        )
+
+    def _scan_fill(self, method: str, limit: int) -> DataFrameOrSeries:
+        # TODO: can this be more efficient?
+
+        # first, compute the outer join between `self.obj` and the `bin_labels`
+        # to get the sampling "gaps":
+        upsampled = self.obj._align_to_index(
+            self.grouping.bin_labels,
+            how="outer",
+            sort=True,
+            allow_non_unique=True,
+        )
+
+        # fill the gaps:
+        filled = upsampled.fillna(method=method)
+
+        # filter the result to only include the values corresponding
+        # to the bin labels:
+        return filled._align_to_index(
+            self.grouping.bin_labels,
+            how="right",
+            sort=False,
+            allow_non_unique=True,
+        )
+
+
+class DataFrameResampler(_Resampler, DataFrameGroupBy):
+    pass
+
+
+class SeriesResampler(_Resampler, SeriesGroupBy):
+    pass
+
+
+class _ResampleGrouping(_Grouping):
+
+    bin_labels: cudf.Index
+
+    def _handle_frequency_grouper(self, by):
+        # if `by` is a time frequency grouper, we bin the key column
+        # using bin intervals specified by `by.freq`, then use *that*
+        # as the groupby key
+        import cudf._lib.labeling
```

Comment:
can this be removed or moved to module level?

Reply:
Moved.